### ISSUE TITLE
Investigate native support for stream wrappers for import paths

### ISSUE THREAD

**Issue by stof** [scssphp/scssphp#256]
#214 reports than Gantry 5 overrides the importer-related methods to use their `gantry-theme` stream wrapper to resolve file.
We should look at a way to support such use case through official extension point, i.e. `addImportPath` with either a string or a callable. If that cannot be done directly, we might need to introduce a more powerful `Importer` API to support that.

**Comment by mahagr** [comment]
@stof I've been reworking the code to use callables to get the real file paths and it seems to work so far (except for map files having absolute paths for source files, which I'm still looking at).

I don't think that SCSS compiler needs to add support for streams, they work out of the box even without the callable (I just tested it).

**Comment by stof** [comment]
@mahagr is this using the unmodified compiler or your customized one which changes the handling of imports to use `$this->streamNames` ?

**Comment by mahagr** [comment]
There is no need to have the variable, it was just used to fix some of the issues I ran into with the old code. Gantry 5 code was originally written 5 years ago for the very old compiler. I had tried to keep it somewhat updated, but with this rewrite I am not even passing streams to SCSS compiler as they are resolved in the callable like you suggested.

That said, even the callable is not really needed to handle streams -- `realpath()` works just fine with local PHP streams. 

The only thing I cannot do anymore is to use `$compiler->getParsedFiles()` to determine if recompilation is needed (it has several issues), but I don't think it matters too much as I can likely deal with that inside my callable.

**Comment by stof** [comment]
> The only thing I cannot do anymore is to use `$compiler->getParsedFiles()` to determine if recompilation is needed (it has several issues)

Can you describe these issues ?

> That said, even the callable is not really needed to handle streams -- `realpath()` works just fine with local PHP streams.

Grat news. This means this could be considered resolved...

**Comment by mahagr** [comment]
We already discussed the issue with `getParsedFiles()` remembering the previous compilations and the issue it causes with caching. The second issue is also caching related as the timestamp checks as the paths are real paths and do not take into account overrides, which was the most requested feature for Gantry -- automatically recompile when user overrides the source files in a different location.

I consider this issue closed as well. We can go through the caching issues in a separate issue.

**Comment by stof** [comment]
@mahagr you mean that the stream can resolve to different files ? would calling `filemtime` on the stream_wrapper path work for overrides ? Maybe that would be the need to natively support stream wrappers properly.

**Comment by Cerdic** [comment]
I opened a new issue for the cache invalidation confitions #268

**Comment by mahagr** [comment]
Yes, in our case stream may resolve into a different file. And yes, filemtime would work in this case, it is what I used to do -- that was the main reason why I had extended the Compiler class in the first place. But if you want to have a general support for stream wrapper, using `realpath()` may not be a good idea -- it won't work for example for non-local files or in our case where the url can resolve into a different file. 

In my case caching parsed content of the file using `realpath()` works fine, but for caching the `@import`s it is not working. For imports there are two separate issues: one is stream wrapper support and the other is lookup paths. In both cases the filename can resolve a multiple possible ways.

I took another look into the caching and currently parser cache works great and does not need any changes, but dependency checks are not optimal for detecting all changes.

**Comment by stof** [comment]
@mahagr I meant calling `filemtime('ganty-theme://myfile.scss')` directly, without applying realpath.

**Comment by mahagr** [comment]
That works. 

But for full stream wrapper support, you'd kinda need to do the same everywhere, meaning that stream wrapper URIs should never be treated with `realpath()`. There are also some other small quirks, most notable example is that `dirname()` does not work reliably for every path.

**Comment by stof** [comment]
> There are also some other small quirks

that's why I opened this issue

**Comment by mahagr** [comment]
If we ignore remote streams, all the issues I have seen this far relate to Compiler caching (#268). Using `realpath()` everywhere will cover every other case (with local files) and there is really no need to add logic to handle with streams.

With #268 there is no need to handle streams internally as the filename would be resolved by callable on every call. So that part would have been taken care of. But what if the files are stored for example in a database? Or by using Flysystem?

I think that the best way to allow files to be loaded from any possible source is to create `FileReader` class with something like:
- `getKey(string $url): ?string`
- `getTimestamp(string $key): int`
- `getContents(string $key): string`

This would allow abstract away the filesystem and allow custom imports from any source.

It would not fully fix the Compiler cache but would cover that `filemtime()` call as it would be trivial to add proper support for the stream wrappers -- or about anything else you could think of.

**Comment by stof** [comment]
@mahagr Can you tell me whether the importer system of scssphp 2.0 allows you to solve your use case ?

Given that Sass imports work with URLs and PHP stream wrappers use some paths that look like URLs instead of paths, adding support natively for stream wrappers in the FilesystemImporter looks quite complex to me. I think the code would be a lot cleaner by using a GantryThemeImporter instead, managing the `gantry-theme` scheme (and maybe reusing a FilesystemImporter internally, after resolving the URL to its actual filesystem path, if relevant)
